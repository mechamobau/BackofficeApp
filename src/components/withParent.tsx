import React, {ComponentType} from 'react';

type EmptyObject = Record<string, never>;

/**
 * Função HOC responsável por abstrair a declaração de componentes pai/filho.
 */
function withParent<ParentProps extends any, ChildProps = EmptyObject>(
  Parent: ComponentType<ParentProps>,
) {
  return (Child: ComponentType<ChildProps>) => {
    const Component = (props: Omit<ParentProps, 'children'> & ChildProps) => (
      <Parent {...(props as ParentProps)}>
        <Child {...(props as ChildProps)} />
      </Parent>
    );

    Component.displayName = `withParent(${
      Parent.displayName || Parent.name || 'Unknown'
    })`;

    return Component;
  };
}
export default withParent;
