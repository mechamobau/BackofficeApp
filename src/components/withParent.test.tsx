import React, {FunctionComponent} from 'react';
import {render, within} from '@testing-library/react-native';
import withParent from './withParent';
import {Text, View} from 'react-native';

type WithChildren<T = {}> = T & {children?: React.ReactNode};

const TEST_ID_COMPONENT_A = 'component-a';
function ParentComponent({children}: WithChildren) {
  return <View testID={TEST_ID_COMPONENT_A}>{children}</View>;
}

const TEST_ID_COMPONENT_B = 'component-b';
function ChildComponent() {
  return <Text testID={TEST_ID_COMPONENT_B}>Just a dummy component</Text>;
}

describe('withParent | function | unit test', () => {
  it('should render a component inside another according with argument order', () => {
    const Component = withParent(ParentComponent)(ChildComponent);

    const {getByTestId, toJSON} = render(<Component />);

    const componentA = getByTestId(TEST_ID_COMPONENT_A);
    const componentB = within(componentA).getByTestId(TEST_ID_COMPONENT_B);

    expect(componentA).not.toBeNull();
    expect(componentB).not.toBeNull();

    expect(toJSON()).toMatchSnapshot();
  });

  it('should change display name of component return', () => {
    const Component = withParent(ParentComponent)(
      ChildComponent,
    ) as FunctionComponent;

    const withParentWithoutDisplayName = withParent(
      ({children}: WithChildren) => (
        <View testID={TEST_ID_COMPONENT_A}>{children}</View>
      ),
    );

    const UnknownDisplayNameParentChild =
      withParentWithoutDisplayName(ChildComponent);

    expect(Component.displayName).toBe('withParent(ParentComponent)');
    expect(UnknownDisplayNameParentChild.displayName).toBe(
      'withParent(Unknown)',
    );
  });
});
