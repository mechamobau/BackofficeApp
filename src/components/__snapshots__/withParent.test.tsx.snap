// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withParent | function | unit test should render a component inside another according with argument order 1`] = `
<View
  testID="component-a"
>
  <Text
    testID="component-b"
  >
    Just a dummy component
  </Text>
</View>
`;
